tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -0,0 +1,1414 @@
+   RTFString strClass{className};
+   RTFString strObject{objectName};
+   RTFString strTitle{title};
+
+   RTFKey key(offset, directoryOffset, strClass, strObject, strTitle, 0, 0);
+   key.fObjLen = nbytes;
+   std::uint64_t offsetData = 0;
+   if (nbytes > fCompressor.kMaxSingleBlock) {
+      // Slow path, size of the key is only known after the compressed data are written
+      Write(&strClass, strClass.GetSize(), offset + key.fKeyHeaderSize);
+      Write(&strObject, strObject.GetSize());
+      Write(&strTitle, strTitle.GetSize());
+      offsetData = fFilePos;
+      auto nbytesZip = fCompressor(buffer, nbytes, compression,
+         [this, offsetData](const void *b, size_t n, size_t o) { Write(b, n, offsetData + o); });
+      key.fNbytes = key.fNbytes + nbytesZip;
+      Write(&key, key.fKeyHeaderSize, offset);
+      Write(nullptr, 0, offsetData + nbytesZip);
+   } else {
+      auto nbytesZip = fCompressor(buffer, nbytes, compression);
+      key.fNbytes = key.fNbytes + nbytesZip;
+      Write(&key, key.fKeyHeaderSize, offset);
+      Write(&strClass, strClass.GetSize());
+      Write(&strObject, strObject.GetSize());
+      Write(&strTitle, strTitle.GetSize());
+      offsetData = fFilePos;
+      Write(fCompressor.GetZipBuffer(), nbytesZip);
+   }
+
+   return offsetData;
+}
+
+void ROOT::Experimental::Detail::RPageSinkFile::WriteRecord(
+   const void *buffer, std::size_t nbytes, std::int64_t offset, int compression)
+{
+   fCompressor(buffer, nbytes, compression,
+      [this, offset](const void *b, size_t n, size_t o) { Write(b, n, offset + o); });
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkFile::WriteRawSkeleton()
+{
+   RRawFileHeader rawHeader;
+   rawHeader.fCompress = fOptions.GetCompression();
+   Write(&rawHeader, sizeof(rawHeader), 0);
+   RTFString ntupleName{fNTupleName};
+   Write(&ntupleName, ntupleName.GetSize());
+
+   fControlBlock->fSeekNTuple = fFilePos;
+   fControlBlock->fNTuple.fSeekHeader = fFilePos + sizeof(RTFNTuple);
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkFile::WriteTFileSkeleton()
+{
+   RTFString strTFile{"TFile"};
+   RTFString strFileName{fFileName};
+   RTFString strTList{"TList"};
+   RTFString strStreamerInfo{"StreamerInfo"};
+   RTFString strStreamerTitle{"Doubly linked list"};
+   RTFString strRNTupleClass{"ROOT::Experimental::RNTuple"};
+   RTFString strRNTupleName{fNTupleName};
+   RTFString strEmpty;
+
+   fControlBlock->fHeader = RTFHeader(fOptions.GetCompression());
+
+   // First record of the file: the TFile object at offset 100
+   RTFFile fileRoot;
+   RTFKey keyRoot(100, 0, strTFile, strFileName, strEmpty,
+                  fileRoot.GetSize() + strFileName.GetSize() + strEmpty.GetSize());
+   std::uint32_t nbytesName = keyRoot.fKeyLen + strFileName.GetSize() + 1;
+   fileRoot.fNBytesName = nbytesName;
+   fControlBlock->fHeader.SetNbytesName(nbytesName);
+
+   // Second record: the compressed StreamerInfo with the description of the RNTuple class
```

Comment:
Odd place to put it (probably I am not following the code completely enough).  The StreamerInfo list content (for an arbitrary file) is both of variable size and only known after all the data is written (since it is essentially the list of data type written in this file).

Reply:
This only works because in this context we ensure that the TFile contains exactly a single ntuple.  I added a corresponding comment in the RMiniFile class.